protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,509 @@
+        self.inner_transport.address_translation(server, observed)
+    }
+}
+
+#[derive(Default)]
+struct RelayedMultiaddr {
+    relay_peer_id: Option<PeerId>,
+    relay_addr: Option<Multiaddr>,
+    dst_peer_id: Option<PeerId>,
+    dst_addr: Option<Multiaddr>,
+}
+
+/// Parse a [`Multiaddr`] containing a [`Protocol::P2pCircuit`].
+///
+/// Returns `Ok(Err(provided_addr))` when passed address contains no [`Protocol::P2pCircuit`].
+///
+/// Returns `Err(_)` when address is malformed.
+fn parse_relayed_multiaddr(
+    addr: Multiaddr,
+) -> Result<Result<RelayedMultiaddr, Multiaddr>, RelayError> {
+    if !addr.iter().any(|p| matches!(p, Protocol::P2pCircuit)) {
+        return Ok(Err(addr));
+    }
+
+    let mut relayed_multiaddr = RelayedMultiaddr::default();
+
+    let mut before_circuit = true;
+    for protocol in addr.into_iter() {
+        match protocol {
+            Protocol::P2pCircuit => {
+                if before_circuit {
+                    before_circuit = false;
+                } else {
+                    return Err(RelayError::MultipleCircuitRelayProtocolsUnsupported);
+                }
+            }
+            Protocol::P2p(hash) => {
+                let peer_id = PeerId::from_multihash(hash).map_err(|_| RelayError::InvalidHash)?;
+
+                if before_circuit {
+                    if relayed_multiaddr.relay_peer_id.is_some() {
+                        return Err(RelayError::MalformedMultiaddr);
+                    }
+                    relayed_multiaddr.relay_peer_id = Some(peer_id)
+                } else {
+                    if relayed_multiaddr.dst_peer_id.is_some() {
+                        return Err(RelayError::MalformedMultiaddr);
+                    }
+                    relayed_multiaddr.dst_peer_id = Some(peer_id)
+                }
+            }
+            p => {
+                if before_circuit {
+                    relayed_multiaddr
+                        .relay_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                } else {
+                    relayed_multiaddr
+                        .dst_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                }
+            }
+        }
+    }
+
+    Ok(Ok(relayed_multiaddr))
+}
+
+#[pin_project(project = RelayListenerProj)]
+pub enum RelayListener<T: Transport> {
+    Inner(#[pin] <T as Transport>::Listener),
+    Relayed {
+        queued_new_addresses: VecDeque<Multiaddr>,
+        from_behaviour: mpsc::Receiver<ToListenerMsg>,
+        msg_to_behaviour: Option<BoxFuture<'static, Result<(), mpsc::SendError>>>,
+    },
+}
+
+impl<T: Transport> Stream for RelayListener<T> {
+    type Item = Result<
+        ListenerEvent<RelayedListenerUpgrade<T>, EitherError<<T as Transport>::Error, RelayError>>,
+        EitherError<<T as Transport>::Error, RelayError>,
+    >;
```

Comment:
This Result type is a bit weird with it's errors. Why is it returning a Result and not
```
type Item = ListenerEvent<RelayedListenerUpgrade<T>, EitherError<T::Error, RelayError>>
```
Alternatively if we would like to have a result here, why not remove the `Error` variant from `ListenerEvent`? From what I have seen also in other transports, most return a `Result` anyway so there is no need for it?

Reply:
I think it's because this is how the trait is defined: https://github.com/libp2p/rust-libp2p/blob/master/core/src/transport.rs#L95-L99 to signal a fatal vs non-fatal error.

> /// If this stream produces an error, it is considered fatal and the listener is killed. It
   /// is possible to report non-fatal errors by producing a [`ListenerEvent::Error`].

Reply:
Okay yes that makes sense. Thought I still think it would clean the type a bit if this would already be represented in the `ListenerEvent` like this:
```rust
pub enum ListenerEvent<TUpgr, TErr> {
   ...
   NonFatalError(TErr),
   FatalError(TErr),
}
```
On the other hand I guess a `Result` emphasizes better that it's a fatal error. 

Either way it's not really a concern of this PR.

Reply:
Just something to note, `type Item = Result<T, E>` matches the signature of [`TryStream`](https://docs.rs/futures/0.3.18/futures/stream/trait.TryStream.html).

> Either way it's not really a concern of this PR.

Agreed. Still open for discussing this further in case you are interested in driving it.